java/runtime/src/main/java/org/ray/runtime/serializer/MessagePackSerializer.java

```diff
@@ -0,0 +1,245 @@
+package org.ray.runtime.serializer;
+
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.lang.reflect.Array;
+import java.lang.reflect.Method;
+import java.math.BigInteger;
+import org.msgpack.core.MessageBufferPacker;
+import org.msgpack.core.MessagePack;
+import org.msgpack.core.MessagePacker;
+import org.msgpack.core.MessageUnpacker;
+import org.msgpack.value.ArrayValue;
+import org.msgpack.value.ExtensionValue;
+import org.msgpack.value.IntegerValue;
+import org.msgpack.value.Value;
+
+// We can't pack List / Map by MessagePack, because we don't know the type class when unpacking.
+public class MessagePackSerializer {
+
+  private static final byte CROSS_LANGUAGE_TYPE_EXTENSION_ID = 100;
+  private static final byte LANGUAGE_SPECIFIC_TYPE_EXTENSION_ID = 101;
+  // MessagePack length is an int takes up to 9 bytes.
+  // https://github.com/msgpack/msgpack/blob/master/spec.md#int-format-family
+  private static final int MESSAGE_PACK_OFFSET = 9;
+
+  interface JavaSerializer {
+
+    void serialize(Object object, MessagePacker packer) throws IOException;
+  }
+
+  interface JavaDeserializer {
+
+    Object deserialize(ExtensionValue v);
+  }
+
+
+  private static void pack(Object object, MessagePacker packer, JavaSerializer javaSerializer)
+      throws IOException {
+    if (object == null) {
```

Comment:
I think maybe we can define a list of cross-language-compatible types. So the code structure could be cleaner. And it's easier to see which types can be cross-language serialized.

For example, we can define a class that has:
1) the Java class object.
2) the MessagePack `ValueType`.
3) a predict function that determines if an object belongs to this type.
4) `pack` and `unpack` functions.

Reply:
It will be cleaner than lots of if else. I don't know how much it impacts performance.